src/Core/Content/Cms/Subscriber/CmsVersionMergeSubscriber.php

```diff
@@ -0,0 +1,54 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Content\Cms\Subscriber;
+
+use Shopware\Core\Framework\DataAbstractionLayer\Event\BeforeVersionMergeEvent;
+use Shopware\Core\Framework\Log\Package;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * @internal
+ */
+#[Package('buyers-experience')]
+class CmsVersionMergeSubscriber implements EventSubscriberInterface
+{
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            BeforeVersionMergeEvent::class => 'onBeforeVersionMerge',
+        ];
+    }
+
+    public function onBeforeVersionMerge(BeforeVersionMergeEvent $event): void
+    {
+        $writes = &$event->getWrites();
+
+        if (empty($writes['delete']['cms_block'])) {
+            return;
+        }
+
+        $deletedBlocks = [];
+        foreach ($writes['delete']['cms_block'] as $deletedBlock) {
+            $blockId = $deletedBlock['id'] ?? null;
+            $blockVersionId = $deletedBlock['versionId'] ?? null;
+
+            if ($blockId && $blockVersionId) {
+                $deletedBlocks[$blockId][$blockVersionId] = true;
+            }
+        }
+
+        foreach (['insert', 'update'] as $operation) {
+            if (!empty($writes[$operation]['cms_slot'])) {
+                $writes[$operation]['cms_slot'] = array_values(array_filter(
```

Comment:
Imo this is quite nested and hard to read.

To move the filter part into own function could help and maybe add a comment before that we did not forget about what it does in the future:

```
// Filter slots based on deleted blocks
    foreach (['insert', 'update'] as $operation) {
        if (empty($writes[$operation]['cms_slot'])) {
            continue;
        }

        $writes[$operation]['cms_slot'] = $this->filterSlots($writes[$operation]['cms_slot'], $deletedBlocks);
    }
```

And `filterSlots` includes the inner logic.